### PR TITLE
Add STM32G0B/C support

### LINKED ISSUE #1035: STM32G0B1 currently black listed
The G0 B/C line of devices are currently blacklisted in the repository because of the problem with the shared interrupt lines.
My question would be: What is an acceptable way to support these devices?

1. Quick and dirty patching of the header to have the "old/common" names of the ISR routines as alias.
Pro: easily accomplished
Cons: only 1 of the multiplexed peripherals can run at the same e.g. either I2C2 or I2C3 can be used not both.
2. add support for multiplexed ISRs to I2C, SPI and CAN
Pro: all of the hardware is supported
Cons: quiet complex, lot of work.

So basically what i want to know is: If a patch for Option 1 has a chance of being accepted or would you need Option 2?

### PR DESCRIPTION
**Pull request by chris-durand**
- [x] Enable devices
- [x] Support G0B/C shared interrupt handlers
  - [x] I2C
  - [x] ~~FDCAN~~, unsupported and deactivated on G0B/C for now
  - [x] SPI
- [x] Fix DMA driver (fixed by [device data correction](https://github.com/modm-io/modm-devices/pull/100) in modm-devices)

Not tested yet.

### CODE REVIEW

File: src/modm/platform/i2c/stm32-extended/module.lb [hunk 1]
```diff
@@ -10,6 +11,23 @@
...
+from collections import defaultdict
+import re
+
+global_properties = {}
+
+def get_shared_irqs(device):
+    irq_re = re.compile("I2C\d(_\d)+")
+    shared_irqs = [v["name"] for v in device.get_driver("core")["vector"]]
+    shared_irqs = [v for v in shared_irqs if irq_re.fullmatch(v)]
+    shared_irq_map = defaultdict(list)
```
Comment (chris-durand) [inline src/modm/platform/i2c/stm32-extended/module.lb]:
```suggestion
    shared_irq_map = {}
```

File: src/modm/platform/i2c/stm32-extended/module.lb [hunk 2]
```diff
@@ -64,12 +90,18 @@ def prepare(module, options):
...
+    global_properties["shared_irqs"] = defaultdict(list)
+
     for instance in listify(device.get_driver("i2c")["instance"]):
         module.add_submodule(Instance(int(instance)))
 
     return True
 
 def build(env):
+    device = env[":target"]
+
```
Comment (chris-durand) [inline src/modm/platform/i2c/stm32-extended/module.lb]:
```suggestion
```

File: src/modm/platform/i2c/stm32-extended/module.lb [hunk 3]
```diff
@@ -64,12 +90,18 @@ def prepare(module, options):
...
+    env.substitutions.update(global_properties)
     env.outbasepath = "modm/src/modm/platform/i2c"
-    env.copy("i2c_timing_calculator.hpp")
 
+    env.copy("i2c_timing_calculator.hpp")
+    env.template("i2c_shared_irqs.cpp.in")
```
Comment (chris-durand) [inline src/modm/platform/i2c/stm32-extended/module.lb]:
```suggestion
    if len(global_properties["shared_irqs"]) > 0:
        env.template("i2c_shared_irqs.cpp.in")
```

File: src/modm/platform/i2c/stm32-extended/i2c_shared_irqs.cpp.in [hunk 4]
```diff
@@ -0,0 +1,25 @@
+/*
```
Comment (klsc-zeat) [inline src/modm/platform/i2c/stm32-extended/i2c_shared_irqs.cpp.in:1]: This file will generated empty if you only use I2C1 -> Is there an option to not generate at all if not needed?

File: src/modm/platform/i2c/stm32-extended/i2c_master.cpp.in [hunk 5]
```diff
@@ -323,15 +324,18 @@ namespace
...
+%% if shared_irq
+void
+modm::platform::I2cMaster{{ id }}::irq()
+%% elif single_interrupt
 MODM_ISR(I2C{{ id }})
 %% else
 MODM_ISR(I2C{{ id }}_EV)
 %% endif
 {
 	DEBUG_STREAM("\n=== IRQ ===");
```
Comment (klsc-zeat) [inline src/modm/platform/i2c/stm32-extended/i2c_master.cpp.in:336]: a few lines late CR1 and ICR are always modified -> in the shared case we should probably check if we have a reason to modify anything:
```
uint16_t isr = I2C{{ id }}->ISR;
if (isr & ?MASK? == 0)
{
    return;
}
```

File: src/modm/platform/i2c/stm32-extended/i2c_shared_irqs.cpp.in (hunk 4 above)
Comment (chris-durand) [reply src/modm/platform/i2c/stm32-extended/i2c_shared_irqs.cpp.in:1]: See the comment above. The file won't be generated anymore, if there are no shared interrupts with that change.

File: src/modm/platform/i2c/stm32-extended/i2c_master.cpp.in (hunk 5 above)
Comment (chris-durand) [reply src/modm/platform/i2c/stm32-extended/i2c_master.cpp.in:336]: Thanks for pointing that out. I didn't look too closely at the driver yesterday. We have to return early if no interrupt flag that is enabled is set.

That could work:
```
const uint16_t isr = I2C{{ id }}->ISR;
%% if shared_irq
const auto cr1 = I2C{{ id }}->CR1;
const auto mask = (cr1 & (I2C_CR1_STOPIE | I2C_CR1_RXIE | I2C_CR1_TXIE))
    | ((cr1 & I2C_CR1_TCIE) ? (I2C_ISR_TCR | I2C_ISR_TC) : 0u); 

static_assert(I2C_CR1_STOPIE == I2C_ISR_STOPF);
static_assert(I2C_CR1_TXIE == I2C_ISR_TXIS);
static_assert(I2C_CR1_RXIE == I2C_ISR_RXNE);

if ((isr & mask) == 0) {
    return;
}
%% endif
```

What do you think?

**Comment by chris-durand** [comment]
@klsc-zeat Do you require FDCAN support on STM32G0B/C? I have disabled the driver on those devices for now.

File: src/modm/platform/i2c/stm32-extended/i2c_master.cpp.in (hunk 5 above)
Comment (klsc-zeat) [reply src/modm/platform/i2c/stm32-extended/i2c_master.cpp.in:336]: I can test this next week -> I need to build a setup that uses both I2C2 and I2C3 first.

**Comment by klsc-zeat** [comment]
> @klsc-zeat Do you require FDCAN support on STM32G0B/C? I have disabled the driver on those devices for now.

Unfortunately yes but only FDCAN1. But i think i will manage that with the examples from the branch. I will post a PR but it will probably take me some time.

File: src/modm/platform/i2c/stm32-extended/i2c_master.cpp.in (hunk 5 above)
Comment (klsc-zeat) [reply src/modm/platform/i2c/stm32-extended/i2c_master.cpp.in:336]: Ok managed to test your branch today. I2C2 and I2C3 work fine in parallel on my Nucleo-G0B1RE.
image
P.S. pinging 2 HYT221 here

**Review by klsc-zeat — APPROVED** [review]

**Review by salkinium — APPROVED** [review]
Awesome, this has been on my todo list for a long time!

**Pull request merged**